### ISSUE TITLE
Object ownership with EventNavigator

### ISSUE THREAD

**Issue by litchfld** [alcap-org/AlcapDAQ#31]
Hi Ben,

So I'm trying to implement the event navigator but I'm getting a bit suck with the idea of adding new information.  I'm looking at MakeAnalysedPulses and I suppose the important line is L129:

https://github.com/alcap-org/AlcapDAQ/blob/feature/data_manager/analyzer/rootana/src/TAP_generators/MakeAnalysedPulses.cpp#L129 

Which I would replace with a EventNavigator::Adopt(...) method.  Presumably that should be roughly be:

``` C++
EventNavigator::Adopt(AnalysedPulseList_t userList, DetectorId id){
  AnalysedPulseList_t internal = this->GetThisSomehow(id);  //will return a clean list
  internal.swap(userList);
}
```

And at the end of an event (L131) we do something like:

``` C++
EventNavigator::Save(){
  //save everything
  //...
  for (/* a loop over Ids */){
    AnalysedPulseList_t internal = this->GetThisSomehow(id); //will return a full list
    internal.clear(); // making sure all pointer elements also delete their targets.
  }
}
```

This seems to ensure that the containers are kept until after saving has happened, and cleared shortly after that.  But it does invalidate any iterators that anyone else is holding, and prevent anyone inspecting the pulses after they have been adopted. I think that is inevitable for sanity reasons, but i wanted to check that it doesn't go against any design ideas you had?

**Comment by benkrikler** [comment]
I'm not sure I completely understand what you're saying.  When you say the end of the event and give L131, you mean the end of MakeAnalysedPulses?  Does that mean that after MakeAnalysedPulses finishes processing an event we'd lose access to the TAPs it has produced?  If that's true, that's definitely NOT what the Event Navigator was supposed to do, as it should also be providing access to the 'adopted' lists, for example, to MakeDetectorPulses which will need them.

Other things that strike me in your examples:
1. Your adopt method just takes a list of TAPs as an argument.  Does this list already contain some identifier for the channel that produced these pulses (ie. are you sub-classing a vector<TAP>, not just typedef-ing it, and then adding a channel identifier field)?  If that is just a vector of TAPS, maybe your adopt method could take a channel ID as another argument, which we would be able to use to retrieve the pulses later.
2.  As descriptive a method call like `this->IGetThisSomehow()` is, what is it actually returning?  Surely it should be a pointer or reference to an std::vector<TAP> member of EventNavigator and not a copy? Assuming there are several lists (one per channel) how is it identifying which one to obtain?

With regards to the title of this thread, we should consider the EventNavigator to own the pulses and the pulse lists once the generator has finished creating them and they've been passed to the EventNavigator (presumably by Adopt(...)).  It should be responsible for clearing the lists and deleting the pulses which no one else should do. In the same vein, it would be good if accessors to the adopted pulses generally only returned const pointers or iterators to the pulses so we can prevent subsequent modules from modifying them.  There are obviously use cases where pulses would be modified, but I guess they're in the minority, so I'd make the more obvious function calls only return immutable objects.

**Comment by litchfld** [comment]
Sorry, I suppressed the detectorId when I first wrote that, then edited it back in because i figured it was confusing.

And yes, thinking harder, doing the delete at the end of MakeAnalysedPulses::ProcessEvent is too soon. You want to be able to get const references back until the end of the event. I think the sensible way to do that is to save at that point too.  But the fact that the original vector 'disappears' after adoption is not a problem, right?

PS: 

> 1. As descriptive a method call like this->IGetThisSomehow() is, what is it actually returning? Surely it should be a pointer or reference to an std::vector member of EventNavigator and not a copy? Assuming there are several lists (one per channel) how is it identifying which one to obtain

It's a shorthand for "this is my private implementation detail so let's not worry about it" =)
